doc/rozo.1
.\" Process this file with
.\" groff -man -Tascii rozo.1
.\"
.TH ROZO 1 "April 2015" RozoFS "User Manuals"
.SH NAME
rozo \- RozoFS management tool
.SH SYNOPSIS
.B rozo
[\fB\-d\fP]
\fItopic
\fP[\fIoptions\fP]
\fIcommand
\fP[\fIoptions\fP]
.PP
.B rozo
\fB\-h\fP | \fB\-\-help\fP
[\fItopic\fP [\fIaction\fP]]
.PP
.B rozo
\fB\-V\fP | \fB\-\-version\fP
.PP
.SH DESCRIPTION
.B rozo
manages a
.BR rozofs (7)
platform.
.SS
General options:
.TP
\fB\-d\fP, \fB\-\-debug
set debugging on.
.TP
\fB\-h\fP, \fB\-\-help
display help and exit.
.TP
\fB\-V\fP, \fB\-\-version
display version information and exit.
.SH AGENT TOPIC
.B rozo agent
.I status \fP|\fI start \fP|\fI stop \fP|\fI restart
.SS status
.B rozo agent status
.TP
Get rozo agent status for this node.
.SS start
.B rozo agent start \fP[\fB\-p\fP | \fB\-\-pacemaker\fP [\fIresource\fP] ] [\fB\-l\fP | \fB\-\-listener\fP [\fIlistener\fP...] ]
.TP
\fPStart a rozo agent with the given \fIlistener\fP list.
.TP
\fB\-p\fP, \fB--pacemaker\fP [\fIresource]
Use this cluster resource name when 
.BR exportd (8) 
is managed thru pacemaker (default: exportd-rozofs).
.TP 
\fB\-l\fP, \fB--listener\fP [\fIlistener\fP...]
List of listener. Must be a subset of {exportd,storaged,rozofsmount}.

.SS stop
.B rozo agent stop
.TP
Stop rozo agent for this node.

.SS restart
.B rozo agent restart \fP[\fB\-p\fP | \fB\-\-pacemaker\fP [\fIresource\fP] ] [\fB\-l\fP | \fB\-\-listener\fP [\fIlistener\fP...] ]
.TP
\fPStop any running rozo agent then start it with the given \fIlistener\fP list.
.TP
\fB\-p\fP, \fB--pacemaker\fP [\fIresource]
Use this cluster resource name when 
.BR exportd (8) 
is managed thru pacemaker (default: exportd-rozofs).
.TP 
\fB\-l\fP, \fB--listener\fP [\fIlistener\fP...]
List of listener. Must be a subset of {exportd,storaged,rozofsmount}.


.\""" rozo exportd
.SH EXPORTD TOPIC
.B rozo exportd
.I layout\fP | \fIoption\fP \fB\-E | \-\-exportd\fP \fIhost\fP... 
.PP

.\" rozo exportd layout subcommands
.RS -4
.B EXPORTD LAYOUT SUBCOMMANDS:
.RI

.\" rozo exportd layout get
.SS layout get
.B rozo exportd layout get \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Get the storage \fIlayout\fP for this platform, see \fBexportd(8)\fP,\fBrozofs(7)\fP.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.

.\" rozo exportd layout set
.SS layout set
.B rozo  exportd layout set \fIlayout\fP \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Set the protection configuration for this platform according to the given \fIlayout\fP, see \fBexportd(8)\fP, \fBrozofs(7)\fP. The \fIlayout\fP value must be 0, 1 or 2.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.

.\" rozo exportd option subcommands
.RS -4
.B EXPORTD OPTION SUBCOMMANDS:
.RI

.\" rozo exportd option list
.SS option list
.B rozo exportd option list \fB\-E | \-\-exportd\fP \fIhost\fP 
.TP 
Display all exportd options and theirs values for the exportd \fIhost\fP.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.PP

.\" rozo exportd option get
.SS option get
.B rozo exportd option get \fIoption\fP \fB\-E | \-\-exportd\fP \fIhost\fP 
.TP 
Get the value for a specific \fIoption\fP name from the exportd \fIhost\fP.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.PP

.\" rozo exportd option set
.SS option set
.B rozo exportd option set \fIoption\fP \fIvalue\fP \fB\-E | \-\-exportd\fP \fIhost\fP 
.TP 
Set a \fIvalue\fP for a specific \fIoption\fP name for the exportd \fIhost\fP.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.PP


.SH NODE TOPIC
.B rozo node
.I list\fP | \fIstatus\fP | \fIconfig\fP | \fIstart\fP | \fIstop\fP | \fIrebuild\fP [\fB\-r | \-\-roles\fP \fIrole\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP... 

.\" rozo node list
.SS list
.B rozo node list [\fB\-r | \-\-roles\fP \fIrole\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Display the list of nodes managed by this platform. Nodes are specified with their roles.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP 
\fB\-r\fP, \fB--roles \fP\fI role\fP...
List of roles. Must be a subset of {exportd,storaged,rozofsmount}. Only nodes with these roles will be listed. If not set all nodes with all roles will be displayed.

.\" rozo node status
.SS status
.B rozo node status [\fB\-r | \-\-roles\fP \fIrole\fP...] [\fB\-n | \-\-nodes\fP \fInode\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Display the list of node managed by this platform. Check the status of nodes and their associated RozoFS components. Nodes and roles can be set to filter the output. If none is specified, everything is displayed.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-r\fP, \fB--roles \fP\fIrole\fP... 
List of roles. Must be a subset of {exportd,storaged,rozofsmount}. Only nodes with these roles will be listed. If not set all nodes with all roles will be displayed.
.TP
\fB\-n\fP, \fB--nodes \fP\fInode\fP...
List of node hostnames to be displayed. If not set all nodes of the platform will be displayed.

.\" rozo node config
.SS config
.B rozo node config [\fB\-r | \-\-roles\fP \fIrole\fP...] [\fB\-n | \-\-nodes\fP \fInode\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Display the list of nodes managed by this platform. Check the status of nodes and their associated RozoFS components. Nodes and roles can be set to filter the output. If none is specified, everything is displayed.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-r\fP, \fB--roles \fP\fIrole\fP... 
List of roles. Must be a subset of {exportd,storaged,rozofsmount}. Only nodes with these roles will be listed. If not set all nodes with all roles will be displayed.
.TP
\fB\-n\fP, \fB--nodes \fP\fInode\fP...
List of node hostnames to be displayed. If not set all nodes of the platform will be displayed.

.\" rozo node start
.SS start
.B rozo node start [\fB\-r | \-\-roles\fP \fIrole\fP...] [\fB\-n | \-\-nodes\fP \fInode\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Start node services.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-r\fP, \fB\-\-roles\fP\fI role\fP... 
List of roles. Must be a subset of {exportd,storaged,rozofsmount}. Only nodes with these roles will be listed. If not set all nodes with all roles will be displayed.
.TP
\fB\-n\fP, \fB\-\-nodes\fP\fI node\fP,... 
List of node hostnames to be started. If not set all nodes of the platform will be started.

.\" rozo node stop
.SS stop
.B rozo node stop [\fB\-r |\-\-roles\fP \fIrole\fP...] [\fB\-n | \-\-nodes\fP \fInode\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP
Stop node services.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-r\fP, \fB\-\-roles\fP\fI role\fP... 
List of roles. Must be a subset of {exportd,storaged,rozofsmount}. Only nodes with these roles will be listed. If not set all nodes with all roles will be displayed.
.TP
\fB\-n\fP, \fB\-\-nodes\fP\fI node\fP...
List of node hostnames to be stopped. If not set all nodes of the platform
will be stopped.


.\""" rozo volume
.SH VOLUME TOPIC
.B rozo volume
.I list\fP | \fIstat\fP | \fIget\fP | \fIexpand\fP | \fIremove\fP \fB\-E | \-\-exportd\fP \fIhost\fP... 

.\" rozo volume list
.SS list
.B rozo volume list \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Display the list of the volumes managed by this platform. Volumes contain clusters which contain storages.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.

.\" rozo volume stat
.SS stat
.B rozo volume stat \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Display statistics about the volumes managed by this platform.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.

.\" rozo volume get
.SS get
.B rozo volume get \fIvid\fP... \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Display statistics about the volume according to the given \fIvid(s)\fP.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.

.\" rozo volume expand
.SS expand
.B rozo volume expand \fP[\fB\-v\fP | \fB\-\-vid\fP \fIvid\fP] [\fB\-l\fP | \fB\-\-layout\fP \fIlayout\fP] [\fB\-t\fP | \fB\-\-total\fP \fIdevice-total\fP] [\fB\-m\fP | \fB\-\-mapper\fP \fIdevice-mapper\fP] [\fB\-r\fP | \fB\-\-redundancy\fP \fIdevice-redundancy\fP] \fInode\fP...  \fB\-E \fP| \fB\-\-exportd\fP \fIhost\fP... 
.TP 
Expand a volume relying on the given list of storage \fInode\fP hostnames.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-v\fP, \fB\-\-vid\fP \fIvid\fP
\fBvid\fP of the volume to expand. If not set a new volume will be created.
.TP
\fB\-l\fP, \fB\-\-layout\fP \fIlayout\fP
Specific \fBlayout\fP to use. If not set the default \fBlayout\fP or the \fBlayout\fP of the already defined volume will be used.
.TP
\fB\-t\fP, \fB\-\-total\fP \fIdevice-total\fP
Specific total number of devices to use for each storage. If not set the default value is 1.
.TP
\fB\-m\fP, \fB\-\-mapper\fP \fIdevice-mapper\fP
Specific number of device mapper to use for each storage. If not set the default value is 1.
.TP
\fB\-r\fP, \fB\-\-redundancy\fP \fIdevice-redundancy\fP
Specific number of device redundancy to use for each storage. If not set the default value is 1.

.\" rozo volume remove
.SS remove
.B rozo volume remove \fB\-v | \-\-vid\fP \fIvid\fP... \fB\-E | \-\-exportd\fP \fIhost\fP... 
.TP 
Remove a volume according to the given \fIvid\fP list from this platform. Volume can be removed only if no export is defined on it.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-v\fP, \fB\-\-vid\fP \fIvid\fP...
List of the \fBvid\fP of the volumes to remove.

.\""" rozo export
.SH EXPORT TOPIC
.B rozo export
.I get\fP | \fIcreate\fP | \fIupdate\fP | \fIremove\fP \fB\-E | \-\-exportd\fP \fIhost\fP... 

.\" rozo export get
.SS get
.B rozo export get [\fB\-e\fP | \fB--eids\fP \fIeid\fP...] \fB\-E | \-\-exportd\fP \fIhost\fP 
.TP 
Display the list of export(s) configuration managed by this platform. If option \fB\-e\fP | \fB--eids\fP \fIeid\fP... is not specified, all export(s) configuration are displayed.

.TP
\fB\-e\fP, \fB\-\-eids \fP\fIeid\fP...
List of export(s) identified by \fIeid\fP... to display. If not set all export(s) configuration are displayed.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.

.SS create
.B rozo export create \fIvid\fP \fB\-E\fP | \fB--exportd\fP \fIhost\fP... [\fB\-n\fP | \fP\-\-name\fP\fI name\fP] [\fB\-p\fP | \fB--passwd\fP\fI password\fP] [\fB\-s\fP | \fB--squota\fP\fI squota\fP] [\fB\-a\fP | \fB--hquota\fP\fI hquota\fP]
.TP
Export a new filesystem on the given volume by \fIvid\fP.
.TP
\fB\-E\fP, \fB--exportd \fP\fIhost\fP...
List of running exportd hostname(s). When securing the exportd thanks to a
virtual IP address only the VIP address is declared and when no VIP is used,
the list of "physical" IP addresses of each host securing the
exportd must be given.
.TP
\fB\-n\fP, \fB\-\-name \fP\fIname
The name to give to the new export. If not set a name will be generated (export_x).
.TP
\fB\-p\fP, \fB\-\-passwd \fP\fIpasswd
Password to set.
.TP
\fB\-s\fP, \fB\-\-squota \fP\fIsquota
Soft quota to set. (value [K | M | G | T]) 
.TP
\fB\-a\fP, \fB\-\-hquota \fP\fIhquota
Hard quota to set. (value [K | M | G | T]) 

.SS update

debian/changelog
rozofs (2.0.alpha21) stable; urgency=low

  * (rozofs) See git log for changes

 -- Fizians S.A.S. <devel@rozofs.com>  Mon, 01 Jan 2014 00:00:00 +0100


manager/rozofs/www/static/rozo.js
var exportd = '';


function login() {
	exportd = $('#exportd').val();
	$('#platformTree').treegrid({url:exportd+'/configuration'});
	$('#platformTree').treegrid('reload');
	$('#login').hide();
}

function platformTreeNodesToStartStopArgs() {
	var args = {};
	args.nodes = [];
	selected = $('#platformTree').treegrid('getSelected');
	if (selected.type == 'Storage' || selected.type == 'Mount') {
		args.nodes[0] = {}
		args.nodes[0].host = selected.text;
		args.nodes[0].role = selected.role;
	} else {
		children = $('#platformTree').treegrid('getChildren', selected.id);
		for (var i=0, j=0; i < children.length; i++) {
			n = children[i];
			if (n.type == 'Storage' || n.type == 'Mount') {
				args.nodes[j] = {}
				args.nodes[j].host = n.text;
				args.nodes[j].role = n.role;
				j++;
			}
		}
	}
	return args;
}

function nodesStart(){
	// what the fuck !
	//$.post('/nodes/start', args, function(data, status) {$('#nodesTree').tree('reload');}, 'json');
	//$.messager.progress({title:'Starting Nodes', msg:'Processing, Please Wait'});
	showLoader();
	var args = platformTreeNodesToStartStopArgs();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/nodes/start',
		success: function(e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function nodesStop(){
	// what the fuck !
	//$.post('/nodes/stop', args, function(data, status) {$('#nodesTree').tree('reload');}, 'json');
	//$.messager.progress({title:'Stopping Nodes', msg:'Processing, Please Wait'});
	showLoader();
	var args = platformTreeNodesToStartStopArgs();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/nodes/stop',
		success: function (e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function platformTreeNodesToAddExportArgs() {
	selected = $('#platformTree').treegrid('getSelected');
	parent = $('#platformTree').treegrid('getParent', selected.id);
	return {'vid': parent.vid};
}

function addExport() {
	var args = platformTreeNodesToAddExportArgs();
	showLoader();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/exports/add',
		success: function (e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function platformTreeNodesToRemoveExportArgs() {
	selected = $('#platformTree').treegrid('getSelected');
	return {'eid': selected.eid};
}

function removeExport() {
	var args = platformTreeNodesToRemoveExportArgs();
	showLoader();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/exports/remove',
		success: function (e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function platformTreeNodesToMountExportsArgs() {
	args = {'eids': []};
	selected = $('#platformTree').treegrid('getSelected');
	if (selected.type == "Export") {
		args.eids[0] = selected.eid;
	} else if (selected.type == "Exports") {
		var children = $('#platformTree').treegrid('getChildren', selected.id);
		var j = 0;
		for (var i=0; i < children.length; i++) {
			var n = children[i];
			if (n.type == "Export") {
				args.eids[j++] = n.eid;
			}
		}
	}
	return args;
}

function mountExports() {
	var args = platformTreeNodesToMountExportsArgs();
	showLoader();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/exports/mount',
		success: function (e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function umountExports() {
	var args = platformTreeNodesToMountExportsArgs();
	showLoader();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/exports/umount',
		success: function (e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function expandVolume() {
	var i = 0;
	var args = {'vid': null, 'hosts': []};
	var selected = $('#platformTree').treegrid('getSelected');
	if (selected.type == "Volume") {
		args.vid = selected.vid;
	}
	var safe = $('#platformTree').treegrid('find', 0).layout[2];
	
	$('#detailView').empty();
	$('#detailView').panel('setTitle', 'Expand Volume');
	$('#detailView').append('<div id="cdiv" class="detail-content">');
	$('#cdiv').append('<div id="hdiv">');
	$('#hdiv').append('<label style="margin-right:5px;">Host:</label>');
	$('#hdiv').append('<input id="ii" class="easyui-validatebox" type="text" data-options="required:true" />');
	$('#hdiv').append('<a id="btnSubmit" href="javascript:void(0)" class="easyui-linkbutton" style="margin:5px;">Add</a>');
	$('#btnSubmit').linkbutton({iconCls: 'icon-add'});
    $('#hdiv').append('<a id="btnExpand" href="javascript:void(0)" class="easyui-linkbutton" style="margin:5px;">Expand</a><br/>');
	$('#btnExpand').linkbutton({iconCls: 'icon-ok'});
	$('#btnExpand').linkbutton('disable');
	$('#cdiv').append('</div>');
	$('#detailView').append('</div>');
	$(function(){
		$('#btnSubmit').bind('click', function() {
			args['hosts'][i] = $('#ii').val();
			$('#cdiv').append('<div id="ndiv'+i+'" style="float:left;padding:20px;">');
			$('#ndiv'+i).append('<img src="/static/node.png"/></br>');
			$('#ndiv'+i).append($('#ii').val());
			$('#cdiv').append('</div>');
			$('#ii').val('');
			if (++i >= safe) {
				$('#btnExpand').linkbutton('enable');
			}
		});
    });
    
    $(function(){
		$('#btnExpand').bind('click', function(){
			showLoader();
			$.ajax({
				type: 'POST',
				contentType: 'application/json',
				data: JSON.stringify(args),
				dataType: 'json',
				url: '/'+exportd+'/volumes/expand',
				success: function (e) {
					$('#detailView').empty();
					hideLoader();
					$('#platformTree').treegrid('reload');
				},
				error:  function (e) {
					$('#detailView').empty();
					hideLoader();
					showError(e);
				}
			});
		});
    });
}

function platformTreeNodesToRemoveVolumeArgs() {
	selected = $('#platformTree').treegrid('getSelected');
	return {'vid': selected.vid};
}

function removeVolume() {
	var args = platformTreeNodesToRemoveVolumeArgs();
	showLoader();
	$.ajax({
		type: 'POST',
		contentType: 'application/json',
		data: JSON.stringify(args),
		dataType: 'json',
		url: '/'+exportd+'/volumes/remove',
		success: function (e) {
			hideLoader();
			$('#platformTree').treegrid('reload');
		},
		error:  function (e) {
			hideLoader();
			showError(e);
		}
	});
}

function showError(e) {
	$.messager.alert('Error', e.message, 'error');
}

function showLoader() {
	$('#spinner').show();
}

function hideLoader() {
	$('#spinner').hide();
}

function platformTreeStatusFormatter(value) {
	switch(value) {
		case -1: return "<img src='static/red-bullet.png'/>";
		case 0: return "<img src='static/yellow-bullet.png'/>";
		case 1: return "<img src='static/green-bullet.png'/>";
		default: return "<span>"+value+"</span>";
	}
}

function platformTreeOnContextMenu(e, row) {
	e.preventDefault();
	$('#platformTree').treegrid('select', row.id);
	var node = $('#platformTree').treegrid('getSelected');
	$('#platformTree'+node.type+'ContextMenu').menu('show', {left: e.pageX, top: e.pageY});
}

function platformTreePlatformPopulateDetailView() {
	var nodes = $('#platformTree').treegrid('getSelected')['nodes'];
	$('#detailView').panel('setTitle', 'Platform Overview');
	$('#detailView').append('<div id="pdiv" class="detail-content">');
	var i=0;
	for (var key in nodes) {
		n = nodes[key];
		$('#pdiv').append('<div id="ndiv'+i+'" style="float:left;padding:20px;">');
		$('#ndiv'+i).append('<img src="/static/node.png"/></br>');
		if (n['up'] == true) {
			$('#ndiv'+i).append('<span style="color:#90EE90;">'+key+'</span>');
		} else {
			$('#ndiv'+i).append('<span style="color:#F08080;">'+key+'</span>');
		} 
		for (var j=0; j < n['roles'].length; j++) {
			$('#ndiv'+i).append('<br>');
			$('#ndiv'+i).append(n['roles'][j]);
		}
		$('#pdiv').append('</div>');
		i++;
	}
	$('#detailView').append('</div>');
}

function platformTreeVolumePopulateDetailView() {
	var node = $('#platformTree').treegrid('getSelected');
	var vstat = node['vstat'];
	var capacity = vstat.blocks*vstat.bsize/(1024*1024*1024);
	var free = vstat.bfree*vstat.bsize/(1024*1024*1024);
	var used = capacity - free;
	
	var volumeData = [
			{name: "Free", y: Number(free.toFixed(2)), color: "#90EE90"},
			{name: "Used", y: Number(used.toFixed(2)), color: "#F08080"}
		];
    var clusterData = [];
    var children = $('#platformTree').treegrid('getChildren', node.id);
    // ugly ! but we need to know number of clusters
    var nbClusters = 0;
    for (var i=0; i < children.length; i++) {
		if (children[i].type == 'Cluster') {
			nbClusters++;
		}
	}
	for (var i=0, j=0; i < children.length; i++) {
		var n = children[i];
		if (n.type == 'Cluster') {
			var cstat = n['cstat'];
			var nsize = cstat.size/(1024*1024*1024);
			var nfree = cstat.free/(1024*1024*1024);
			var nused = nsize - nfree;
			clusterData[j] = {name: n.text, y: Number(nfree.toFixed(2)), color: Highcharts.Color("#90EE90").brighten((0.02)*(i+1)).get()};
			clusterData[j+nbClusters] = {name: n.text, y: Number(nused.toFixed(2)), color: Highcharts.Color("#F08080").brighten((0.02)*(i+1)).get()};
			j++;
		}
	}

	$('#detailView').panel('setTitle', 'Cluster id: ' + node['cid']);
	$('#detailView').append('<div id="cdiv" class="detail-content">');
	$('#detailView').append('</div>');
    
	// Create the chart
    $('#cdiv').highcharts({
		chart: {type: 'pie'},
        title: {text: 'Volume Usage'},
        credits: {enabled: false},
        yAxis: {title: {text: 'Total volume usage'}},
        plotOptions: {pie: {shadow: true,center: ['50%', '50%']}},
        tooltip: {valueSuffix: 'GB'},
        series: [{
				name: 'Capacity',
				data: volumeData,
				size: '60%',
				dataLabels: {
					formatter: function() {
						return '<b>'+ this.point.name +': '+ this.y +'GB</b>'
					},
					color: 'white',
					distance: -30
				}
            }, {
                name: 'Capacity',
                data: clusterData,
                size: '80%',
                innerSize: '60%',
                dataLabels: {
                    formatter: function() {
						return '<b>'+ this.point.name +': '+ this.y +'GB</b>'
                    }
                }
            }]
        });
}

function platformTreeClusterPopulateDetailView() {
	var node = $('#platformTree').treegrid('getSelected');
	var cstat = node['cstat'];
	var size = cstat.size/(1024*1024*1024);
	var free = cstat.free/(1024*1024*1024);
	var used = size - free;
	
	var clusterData = [
			{name: "Free", y: Number(free.toFixed(2)), color: "#90EE90"},
			{name: "Used", y: Number(used.toFixed(2)), color: "#F08080"}
		];
    var nodesData = [];
    var children = $('#platformTree').treegrid('getChildren', node.id);
	for (var i=0; i < children.length; i++) {
		var n = children[i];
		var sstat = n['sstat'];
		var nsize = sstat.size/(1024*1024*1024);
		var nfree = sstat.free/(1024*1024*1024);
		var nused = nsize - nfree;
		nodesData[i] = {name: n.text, y: Number(nfree.toFixed(2)), color: Highcharts.Color("#90EE90").brighten((0.02)*(i+1)).get()};
		nodesData[i+children.length] = {name: n.text, y: Number(nused.toFixed(2)), color: Highcharts.Color("#F08080").brighten((0.02)*(i+1)).get()};
	}

	$('#detailView').panel('setTitle', 'Cluster id: ' + node['cid']);
	$('#detailView').append('<div id="cdiv" class="detail-content">');
	$('#detailView').append('</div>');
    
	// Create the chart
    $('#cdiv').highcharts({
		chart: {type: 'pie'},
		credits: {enabled: false},
        title: {text: 'Cluster Usage'},
		yAxis: {title: {text: 'Total volume usage'}},
		plotOptions: {pie: {shadow: true, center: ['50%', '50%']}},
        tooltip: {valueSuffix: 'GB'},
        series: [{
                name: 'Capacity',
                data: clusterData,
                size: '60%',
                dataLabels: {
                    formatter: function() {
						return '<b>'+ this.point.name +': '+ this.y +'GB</b>'
                    },
                    color: 'white',
                    distance: -30
                }
            }, {
                name: 'Capacity',
                data: nodesData,
                size: '80%',
                innerSize: '60%',
                dataLabels: {
                    formatter: function() {
						return '<b>'+ this.point.name +': '+ this.y +'GB</b>'
                    }
                }
            }]
        });
}

function platformTreeExportPopulateDetailView() {
	var node = $('#platformTree').treegrid('getSelected');
	var estat = node['estat'];
	var capacity = estat.blocks*estat.bsize/(1024*1024*1024);
	var free = estat.bfree*estat.bsize/(1024*1024*1024);
	var used = capacity - free;
	var capacityData = [
			{name: "Free", y: Number(free.toFixed(2)), color: "#90EE90"},
			{name: "Used", y: Number(used.toFixed(2)), color: "#F08080"}
		];
	var files = estat.files;
	var ffree = estat.ffree;
	var filesData = [
			{name: "Free", y: Number(ffree.toFixed(2)), color: "#90EE90"},
			{name: "Used", y: Number(files.toFixed(2)), color: "#F08080"}
		];
	
	$('#detailView').panel('setTitle', 'Export id: ' + node['eid']);
	$('#detailView').append('<div id="content" class="detail-content">');
	$('#content').append('<div id="ldiv" class="detail-left"></div><div id="rdiv" class="detail-right"></div>');
	$('#detailView').append('</div>');
	// Create the charts
    $('#ldiv').highcharts({
		chart: {type: 'pie'},
        title: {text: 'Export Capacity Usage'},
        yAxis: {title: {text: 'Total export usage'}},
        plotOptions: {pie: {shadow: true, center: ['50%', '50%']}},
		tooltip: {valueSuffix: 'GB'},
        series: [{
                name: 'Capacity',
                data: capacityData,
                size: '80%',
                dataLabels: {
                    formatter: function() {
						return '<b>'+ this.point.name +': '+ this.y +'GB</b>'
                    },
                    color: 'white',
                    distance: -30
                }
            }]
        });
	$('#rdiv').highcharts({
		chart: {type: 'pie'},
        title: {text: 'Export Files Usage'},
        yAxis: {title: {text: 'Total export usage'}},
        plotOptions: {pie: {shadow: true, center: ['50%', '50%']}},
        series: [{
                name: 'Files',
                data: filesData,
                size: '80%',
                dataLabels: {
                    formatter: function() {
						return '<b>'+ this.point.name +': '+ this.y +'</b>'
                    },
                    color: 'white',
                    distance: -30
                }
            }]
        });
}

function platformTreeOnClickRow(row) {
	$('#detailView').empty();
	var node = $('#platformTree').treegrid('getSelected');
	if (node.type == 'Platform') {
		platformTreePlatformPopulateDetailView();
	} else if (node.type == 'Volume') {
		platformTreeVolumePopulateDetailView();
	} else if (node.type == 'Cluster') {
		platformTreeClusterPopulateDetailView();
	} else if (node.type == 'Export') {
		platformTreeExportPopulateDetailView();
	}
}
